server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/core/JmapRfc8621Configuration.scala

```diff
@@ -43,33 +44,44 @@ object JmapRfc8621Configuration {
+    websocketPrefixString = WEBSOCKET_URL_PREFIX_DEFAULT,
+    maxUploadSize = UPLOAD_LIMIT_DEFAULT)
 
-
-  def from(configuration: Configuration): JmapRfc8621Configuration = {
+  def from(configuration: Configuration): JmapRfc8621Configuration =
     JmapRfc8621Configuration(
       urlPrefixString = Option(configuration.getString(URL_PREFIX_PROPERTY)).getOrElse(URL_PREFIX_DEFAULT),
       websocketPrefixString = Option(configuration.getString(WEBSOCKET_URL_PREFIX_PROPERTY)).getOrElse(WEBSOCKET_URL_PREFIX_DEFAULT),
+      dynamicJmapPrefixResolutionEnabled = configuration.getBoolean(DYNAMIC_JMAP_PREFIX_RESOLUTION_ENABLED_PROPERTY, false),
       maxUploadSize = Option(configuration.getString(UPLOAD_LIMIT_PROPERTY, null))
         .map(Size.parse)
         .map(MaxSizeUpload.of(_).get)
         .getOrElse(UPLOAD_LIMIT_DEFAULT),
       maxTimeoutSeconds = Optional.ofNullable(configuration.getInteger(WEB_PUSH_MAX_TIMEOUT_SECONDS_PROPERTY, null)).map(Integer2int).toScala,
       maxConnections = Optional.ofNullable(configuration.getInteger(WEB_PUSH_MAX_CONNECTIONS_PROPERTY, null)).map(Integer2int).toScala)
-  }
 }
 
-case class JmapRfc8621Configuration(urlPrefixString: String, websocketPrefixString: String,
+case class JmapRfc8621Configuration(urlPrefixString: String,
+                                    websocketPrefixString: String,
+                                    dynamicJmapPrefixResolutionEnabled: Boolean = false,
                                     maxUploadSize: MaxSizeUpload = UPLOAD_LIMIT_DEFAULT,
                                     maxTimeoutSeconds: Option[Int] = None,
                                     maxConnections: Option[Int] = None) {
-  val urlPrefix: URL = new URL(urlPrefixString)
-  val apiUrl: URL = new URL(s"$urlPrefixString/jmap")
-  val downloadUrl: URL = new URL(urlPrefixString + "/download/{accountId}/{blobId}?type={type}&name={name}")
-  val uploadUrl: URL = new URL(s"$urlPrefixString/upload/{accountId}")
-  val eventSourceUrl: URL = new URL(s"$urlPrefixString/eventSource?types={types}&closeAfter={closeafter}&ping={ping}")
-  val webSocketUrl: URI = new URI(s"$websocketPrefixString/jmap/ws")
 
+  def getURLPrefix(urlPrefixRequest: Option[URL]): String = Some(dynamicJmapPrefixResolutionEnabled)
+    .filter(enabled => enabled)
+    .flatMap(_ => urlPrefixRequest.map(_.toString))
+    .getOrElse(urlPrefixString)
+
+  def apiUrl(urlPrefixRequest: Option[URL] ): URL = new URL(s"${getURLPrefix(urlPrefixRequest)}/jmap")
+
+  def downloadUrl(urlPrefixRequest: Option[URL]): URL = new URL(s"${getURLPrefix(urlPrefixRequest)}/download/{accountId}/{blobId}?type={type}&name={name}")
+
+  def uploadUrl(urlPrefixRequest: Option[URL]): URL = new URL(s"${getURLPrefix(urlPrefixRequest)}/upload/{accountId}")
+
+  val webSocketUrl: URI = new URI(s"$websocketPrefixString/jmap/ws")
+  val eventSourceUrl: URL = new URL(s"$urlPrefixString/eventSource?types={types}&closeAfter={closeafter}&ping={ping}")
```

Comment:
Don't forget to resolve `webSocketUrl` and `eventSourceUrl` as well

Reply:
Then we need one more header to set WebSocket url prefix. I'm right?

1. X-JMAP-PREFIX:
- https://domain.tld/oidc
- http://domain.tld/oidc

2. X-JMAP-WEBSOCKET-PREFIX
- wss://domain.tld/oidc
- ws://domain.tld/oidc

?

Reply:
I figure out, 
X-JMAP-PREFIX should have only domain and path, NOT contain schema. 
Then we can use it to replace in both urlPrefix and websocketUrlPrefix

Reply:
the X-JMAP-PREFIX is the same for every endpoint, it's unrelated to urlPrefix and websocketUrlPrefix

Reply:
I think the two header approach is valid.